etc/systemd/system/zfs-scrub@.service.in

```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=zpool scrub on %i
+Documentation=man:zpool-scrub(8)
+Requires=zfs.target
+After=zfs.target
+ConditionPathIsDirectory=/sys/module/zfs
+
+[Service]
+KillSignal=SIGINT
+ExecStart=@sbindir@/zpool scrub %i
```

Comment:
agreed, I'll try to make it a unit condition though, not part of `ExecStart`.
In that case systemd will log that unit was skipped due to condition not satisfied.
it should be possible to re-schedule it sooner, not in a month/week, in that case
not sure how exactly the re-scheduling will work, I'll do some research.

Reply:
if it's condition and not an assert - it will not fail the unit, so it should not alert the monitoring needlessly.
it will log something like that `Jun 01 21:21:26 cerberus systemd[1]: Condition check resulted in Import ZFS pools by cache file being skipped.`

Reply:
AFAIK, there is no way to run a command as a Condition. But I could be wrong.

Reply:
this seems to be a good fit.
```
  ExecCondition=
           Optional commands that are executed before the command(s) in ExecStartPre=. Syntax is the same as for ExecStart=, except
           that multiple command lines are allowed and the commands are executed one after the other, serially.

           The behavior is like an ExecStartPre= and condition check hybrid: when an ExecCondition= command exits with exit code 1
           through 254 (inclusive), the remaining commands are skipped and the unit is not marked as failed. However, if an
           ExecCondition= command exits with 255 or abnormally (e.g. timeout, killed by a signal, etc.), the unit will be considered
           failed (and remaining commands will be skipped). Exit code of 0 or those matching SuccessExitStatus= will continue execution
           to the next command(s)
````

also after thinking a bit I realized it may be worth using ` ExecStop=` with `zpool-scrub -s` or `-p`

or maybe `-w` to scrub process itself.

because `SIGINT` to scrub may actually mean nothing, as it just signals the kernel and exits.


another thing to add is `ConditionACPower` to skip scrubs on battery power only.

Reply:
`ExecCondition` appeared in `systemd-243` according to https://github.com/systemd/systemd/blob/05a2166a068f54d59d68590a81369768c2eb657b/NEWS#L3194

so I guess it's too new for some distros and their respective still-supported versions.

Reply:
So assuming you can invert an `ExecCondition` with `!`, something like this should work:
```
ExecCondition=!/bin/sh -c 'zpool status %i | grep -q "scrub in progress"'
ExecStart=@sbindir@/zpool scrub %i
```

But: 1) As you said, `ExecCondition` is too new. We could write some autoconf check for the systemd version and conditionalize it, but that seems like overkill because... 2) This doesn't really buy us much. We still end up needing to invoke `sh` so we can pipe to `grep`.

`ConditionACPower` seems reasonable.

Regarding `ExecStop`: I think that's reasonable if and only if you can avoid running the stop at shutdown (but see notes below). Otherwise, if I reboot the system, the scrub will stop/pause and will not restart/resume again on boot. Instead, it'll only restart/resume the scrub at the next scheduled interval. On servers where scrubs take days or weeks (and I have such things), that's not good. But if you can avoid systemd running stop on it on shutdown, then I think it'd be fine, as then the `ExecStop` would only happen upon explicit user request. I think pausing the scrub is the safer option, but I'm open to arguments about why `systemctl stop zfs-scrub@pool` should actually stop the scrub. (The obvious one being "stop" is the verb you're using.)

However, for the `systemctl stop` to actually do anything, systemd has to think the service is running. So you'll need to set `RemainAfterExit=true` or you could (but I'm not saying you have to) wire in `zpool wait -t scrub %i` If you use `zpool wait`, the service would stay in an "active (running)" state as long as the scrub is running (well, until the system reboots). This makes the systemd state make more sense. However, I'm not sure how you avoid calling the ExecStop on shutdown.

Reply:
I think `DefaultDependencies=no` will avoid stopping the service on shutdown. Since you already have ordering for startup, those are probably sufficient.

Reply:
Please squash.

nit: Type=simple is the default, so it can be ommitted.
nit: I'd put an `exec` on the scrub command, so the sh process does not need to stick around.

If it is going to wait when it starts the scrub, should it also wait if the scrub is running already?

If so, then this works:

```
[Service]
ExecStart=/bin/sh -c '\
if /sbin/zpool status %i | grep -q "scrub in progress" ; then\
echo "scrub in progress"; exec /sbin/zpool wait -t scrub %i;\
else exec /sbin/zpool scrub -w %i; fi'
```

If not, then this works (which is what you wrote, but with the two nits above addressed and the specific ExecCondition figured out):
```
[Service]
# Use this once systemd >= 243 is widespread:
#ExecCondition=/bin/sh -c '! /sbin/zpool status %i | grep -q "scrub in progress"'
#ExecStart=/sbin/zpool scrub -w %i
ExecStart=/bin/sh -c '\
if /sbin/zpool status %i | grep -q "scrub in progress" ; then\
echo "scrub in progress";\
else exec /sbin/zpool scrub -w %i; fi'
```

Note that the negation needs to happen in the shell, as ExecCondition only works one way.

Another option would be to make the grep non-quiet, so the zpool status message comes through as the status. Here is the second example modified in that way:
```
[Service]
ExecStart=/bin/sh -c '\
if ! /sbin/zpool status %i | grep "scrub in progress" ; then\
exec /sbin/zpool scrub -w %i; fi'
```
That produces output like this:
```
Jul 04 00:31:21 sid systemd[1]: Started zpool scrub on test.
Jul 04 00:31:21 sid sh[75574]:   scan: scrub in progress since Sun Jul  4 00:31:21 2021
Jul 04 00:31:21 sid systemd[1]: zfs-scrub@test.service: Succeeded.
```

In testing this, I prefer not using ExecCondition. Having log entries like this make it seem like something went _wrong_, as opposed to the fact that the scrub is _already_ occurring:
```
Jul 04 00:36:35 sid systemd[1]: zfs-scrub@test.service: Control process exited, code=exited, status=1/FAILURE
Jul 04 00:36:35 sid systemd[1]: zfs-scrub@test.service: Skipped due to 'exec-condition'.
```

I think I have a mild preference in favor of always waiting on the scrub. That just feels more consistent.

I kinda like showing the "scrub in progress" line.

So I end up with this:
```
[Service]
ExecStart=/bin/sh -c '\
if /sbin/zpool status %i | grep "scrub in progress" ; then\
exec /sbin/zpool wait -t scrub %i;\
else exec /sbin/zpool scrub -w %i; fi'
```

What do you think?

To be clear, I've tested all of these versions.

Longer-term, it'd be nice if `zpool wait` and `zpool scrub` could support `sd_notify()` to periodically report status information to systemd. But that's well out of the scope of this PR.